diff -urN xt_tls-3636cd5.orig/src/Makefile xt_tls-3636cd5.patch/src/Makefile
--- xt_tls-3636cd5.orig/src/Makefile	2018-11-13 21:49:08.000000000 +0000
+++ xt_tls-3636cd5.patch/src/Makefile	2018-11-20 01:10:27.115008047 +0000
@@ -6,13 +6,13 @@
 PWD := $(shell pwd)
 VERSION := $(shell git rev-parse HEAD 2>/dev/null)
 default:
-	    $(MAKE) -C $(KDIR) SUBDIRS=$(PWD) modules
+	    $(MAKE) -C $(KERNEL_DIR) SUBDIRS=$(PWD) modules
 
 debug:
-	    $(MAKE) -C $(KDIR) SUBDIRS=$(PWD) CFLAGS_MODULE="-DXT_TLS_DEBUG" modules
+	    $(MAKE) -C $(KERNEL_DIR) SUBDIRS=$(PWD) CFLAGS_MODULE="-DXT_TLS_DEBUG" modules
 
 install:
-	    $(MAKE) -C $(KDIR) SUBDIRS=$(PWD) INSTALL_MOD_PATH=$(DESTDIR) modules_install
+	    $(MAKE) -C $(KERNEL_DIR) SUBDIRS=$(PWD) INSTALL_MOD_PATH=$(DESTDIR) modules_install
 	    depmod -a
 
 clean:
Binary files xt_tls-3636cd5.orig/src/.Makefile.swp and xt_tls-3636cd5.patch/src/.Makefile.swp differ
